java/kotlin-extractor/deps.bzl

```diff
@@ -0,0 +1,126 @@
+
+    # for some reason rules_kotlin warns about these jars missing, this is to silence those warnings
+    repository_ctx.file("empty.zip", _empty_zip)
+    for jar in (
+        "annotations-13.0.jar",
+        "kotlin-stdlib.jar",
+        "kotlin-reflect.jar",
+        "kotlin-script-runtime.jar",
+        "trove4j.jar",
+    ):
+        repository_ctx.symlink("empty.zip", jar)
+    repository_ctx.file("BUILD.bazel", _kotlin_dep_build.format(name = name))
+
+_kotlin_dep = repository_rule(
+    implementation = _kotlin_dep_impl,
+)
+
+def _walk(dir):
+    res = []
+    next_dirs = [dir]
+
+    # loops must be bounded in starlark
+    for i in range(100):
+        current_dirs = next_dirs
+        next_dirs = []
+        for d in current_dirs:
+            children = d.readdir()
+            next_dirs.extend([c for c in children if c.is_dir])
+            res.extend([c for c in children if not c.is_dir])
+        if not next_dirs:
+            break
+    return res
+
+def _embeddable_source_impl(repository_ctx):
+    src_dir = repository_ctx.path(Label("//java/kotlin-extractor:src"))
+    for src in _walk(src_dir):
+        contents = repository_ctx.read(src)
+        contents = contents.replace(
+            "import com.intellij",
+            "import org.jetbrains.kotlin.com.intellij",
+        )
+        repository_ctx.file(str(src).replace(str(src_dir), "src"), contents)
+    repository_ctx.symlink(
+        Label("//java/kotlin-extractor:generate_dbscheme.py"),
+        "generate_dbscheme.py",
+    )
+    repository_ctx.symlink(
+        Label("//java/kotlin-extractor:BUILD.bazel"),
+        "BUILD.bazel",
+    )
+
+_embeddable_source = repository_rule(implementation = _embeddable_source_impl)
+
+def _get_default_version(repository_ctx):
+    default_version = repository_ctx.getenv("CODEQL_KOTLIN_SINGLE_VERSION")
+    if default_version:
+        if default_version not in VERSIONS:
+            fail("overriding CODEQL_KOTLIN_SINGLE_VERSION=%s not known, must be one of:\n  %s" %
+                 (default_version, " ".join(VERSIONS)))
+        return default_version
+    kotlinc = repository_ctx.which("kotlinc")
+    if not kotlinc:
+        return DEFAULT_FALLBACK_VERSION
+    res = repository_ctx.execute([kotlinc, "-version"])
+    if not res:
+        fail("kotlinc -version failed: %s" % res.stderr)
+    out = res.stderr.split(" ")
+    if len(out) < 3:
+        fail("malformed kotlinc -version output: %s" % res.stdout)
+    host_version = out[2]
+    for version in reversed(VERSIONS):
+        if version_less(version, host_version) or version == host_version:
+            return version
+    fail("no relevant version found for host version %s among:\n  %s" % (host_version, " ".join(VERSIONS)))
+
+def _defaults_impl(repository_ctx):
+    default_version = _get_default_version(repository_ctx)
+    default_variant = "standalone"
+    if repository_ctx.getenv("CODEQL_KOTLIN_SINGLE_VERSION_EMBEDDABLE") in ("true", "1"):
```

Comment:
Nit: Maybe use https://bazel.build/rules/lib/globals/all.html#bool instead?

Reply:
Ah, it doesn't interpret stringy things properly, then nvm!

Reply:
yeah, it's the equivalent of python's `bool` type, so it models value truthyness rather than stringy booleans.